File: LeetCode-Java/Flatten Binary Tree to Linked List.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void flatten(TreeNode root) {
        Stack<TreeNode> nodeStack = new Stack<TreeNode>();
        if(root == null) return;
        nodeStack.push(root);
        while(!nodeStack.empty())
        {
            TreeNode tempNode = nodeStack.pop();
            if(tempNode.right != null) nodeStack.push(tempNode.right);
            if(tempNode.left != null) nodeStack.push(tempNode.left);
            if(nodeStack.empty()) tempNode.right = null;
            else tempNode.right = nodeStack.peek();
            tempNode.left = null;
        }
        return;
    }
}

File: Leetcode-cpp/flatten_binary_tree_to_linked_list.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode* root) {
        flattenHelper(root);
    }
    
    TreeNode* flattenHelper(TreeNode* root) {
        if (root == nullptr || (root->left == nullptr && root->right == nullptr)) return root;
        TreeNode* rChild = root->right;
        if (root->left != nullptr) {
            root->right = root->left;
            root->left = nullptr;
            TreeNode* lResult = flattenHelper(root->right);
            if (rChild == nullptr) return lResult;
            lResult->right = rChild;
        } 
        return flattenHelper(rChild);
    }
};